core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -81,8 +79,8 @@ public ParseState iterate() {
         return newOffset.compareTo(ZERO) >= 0 ? Optional.of(new ParseState(order, source, newOffset, iterations)) : Optional.empty();
     }
 
-    public ParseState source(final ValueExpression dataExpression, final int index, final ParseState parseState, final Encoding encoding) {
-        return new ParseState(order, new DataExpressionSource(dataExpression, index, parseState, encoding), ZERO, iterations);
+    public ParseState withSource(final Source source) {
```

Comment:
What code smell did this resolve? Not sure if this is a good idea, since this changes the public API. Maybe I would keep the previous method and deprecate this with a reference to this new method?

Reply:
This was the related smell: https://sonarcloud.io/project/issues?id=parsingdata_metal&issues=AWepfYZ8D1JUB5M3GqDE&open=AWepfYZ8D1JUB5M3GqDE

It specifically mentions the name of the method as a problem since it clashes with the name of a field. So then when I figured this would change the public API I also had a look at the arguments. It turns out that the arguments simply get passed directly to the `DataExpressionSource`-constructor. And since the method is called `source` (or `withSource` in the new situation) is makes more sense to have the client pass in a valid `Source` instead of a list of arguments that _only_ map to one single constructor.

Additionally, I think it's not too big of a problem that this changes the public API because (1) the next milestone is a major release and (2) it's not something external users will actually invoke. The method is only used by the implementation of `Tie` currently. The other method that does a similar thing with `offset` is called `seek()`, but there I think that's a good idea because it's a generally accepted and broadly used term for `withOffset()`.